src/lib.rs

```diff
@@ -247,50 +254,129 @@ impl<'a, I, T> core::clone::Clone for Menu<'a, I, T> {
     }
 }
 
-impl<'a, I, T> Runner<'a, I, T>
+#[derive(Clone)]
+enum PromptIterState {
+    Newline,
+    Menu(usize),
+    Arrow,
+    Done,
+}
+
+struct PromptIter<'a, I, T> {
+    menu_mgr: &'a MenuManager<'a, I, T>,
+    state: PromptIterState,
+}
+
+impl<'a, I, T> Clone for PromptIter<'a, I, T> {
+    fn clone(&self) -> Self {
+        Self {
+            menu_mgr: self.menu_mgr,
+            state: self.state.clone(),
+        }
+    }
+}
+
+impl<'a, I, T> PromptIter<'a, I, T> {
+    fn new(menu_mgr: &'a MenuManager<'a, I, T>, newline: bool) -> Self {
+        let state = if newline {
+            PromptIterState::Newline
+        } else {
+            PromptIterState::Menu(0)
+        };
+        Self { menu_mgr, state }
+    }
+}
+
+impl<'a, I, T> Iterator for PromptIter<'a, I, T> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.state {
+                PromptIterState::Newline => {
+                    self.state = PromptIterState::Menu(0);
+                    break Some("\n");
+                }
+                PromptIterState::Menu(i) => {
+                    if i >= self.menu_mgr.depth() {
+                        self.state = PromptIterState::Arrow;
+                    } else {
+                        let menu = self.menu_mgr.get_menu(Some(i));
+                        self.state = PromptIterState::Menu(i + 1);
+                        break Some(menu.label);
+                    }
+                }
+                PromptIterState::Arrow => {
+                    self.state = PromptIterState::Done;
+                    break Some("> ");
+                }
+                PromptIterState::Done => break None,
+            }
+        }
+    }
+}
+
+impl<'a, I, T, B: ?Sized> Runner<'a, I, T, B>
 where
-    I: core::fmt::Write,
+    I: embedded_io::Write,
 {
     /// Create a new `Runner`. You need to supply a top-level menu, and a
     /// buffer that the `Runner` can use. Feel free to pass anything as the
     /// `context` type - the only requirement is that the `Runner` can
     /// `write!` to the context, which it will do for all text output.
-    pub fn new(
-        menu: Menu<'a, I, T>,
-        buffer: &'a mut [u8],
-        mut interface: I,
-        context: &mut T,
-    ) -> Self {
+    pub fn new(buffer: &'a mut B, menu: Menu<'a, I, T>, mut interface: I, context: &mut T) -> Self {
```

Comment:
For consistency we should probably swap the order of these args back as they were.